### ISSUE TITLE
Duplicate detection for vaccinations when merging cases [~5]

### ISSUE THREAD

**Issue by SahaLinaPrueger** [hzi-braunschweig/SORMAS-Project#9299]
### Problem Description

While performing case and contact merging (each case/contact has his own person), here an example what happens with the immunizations/vaccinations:
![Merging](https://user-images.githubusercontent.com/100844689/170082797-c0a8b8f7-8049-4279-a1fe-1a604ee63e39.png)

- non-leading immunization with the person is not discarded 
- vaccinations are added to remaining immunizations without comparing whether they subsequently exist twice. This can complicate evaluations of the number of vaccinations per case/contact due to duplicates.
- Person duplicates are the main issue because immunizations and therefore vaccinations are not duplicated for one person.

### Proposed Change

- Currently, vaccinations are added to remaining immunizations.
    - Merge vaccinations rather than just adding them in the following way:
        - IF entries do not have vaccinationDate AND vaccineName or vaccinationDate OR vaccineName, simply add the vaccination to the remaining immunization
        - IF entries have a matching vaccinationDate AND vaccineName, do not create an additional entry
transfer any additional fields that are empty in the remaining vaccination to the leading vaccination

**Comment by SahaLinaPrueger** [comment]
@JaquM-HZI

**Comment by MateStrysewske** [comment]
@SahaLinaPrueger @kwa20 I'm fine with this, just wanted to remind you that the current behaviour was specifically requested in a joint refinement session because the person of the discarded case could be connected to other entities (e.g. other cases, or contacts), and if we discard the immunization related to the person of the discarded case, that can lead to some kind of data loss (i.e. the data is still there, but not connected to this person instance anymore, and thus won't be shown when looking at data connecting to the person). This can only be solved by implementing the feature to merge duplicate persons that's in our backlog, but has not been prioritised yet.

Long story short: I'd like you to confirm that you're aware of this side effect and are fine with the possible consequences.

**Comment by kwa20** [comment]
@MateStrysewske We've discussed some of the points yesterday, maybe we can use the next refinement meeting to go through the details?

**Comment by roxanamlendea** [comment]
I'm reopening the ticket due to a bug. Tested on local machine version 1.73.0-SNAPSHOT (d0284c9).
Reduced immunization enabled, after two Cases are merged and so is the Vaccination (entries have a matching vaccinationDate AND vaccineName), when I download a detailed export, the Vaccination related fields are empty.

**Comment by MateStrysewske** [comment]
@kwa20 The reason for the bug discovered by @roxanamlendea is that when filling the export list, we simply look at the most recent IMMUNIZATION and then are collecting the vaccination data from it. When two cases are merged with e.g. only one vaccination that is detected as a duplicate, the most recent immunization of the preserved case (which is created during the merging process) does not have any vaccinations however, which is why there's no information displayed in the export.

I guess it would probably make more sense to consider ALL immunizations of the respective disease, and then pick the earliest and latest vaccination from all vaccinations of all those immunizations. Do you agree, or do you see any potential problems with that approach?

**Comment by kwa20** [comment]
@MateStrysewske I think that would be a good approach. Independent of the number of immunizations, all are associated to the same disease and therefore all vaccinations would be shown in the case. All of them should then also be considered for the export.

**Comment by roxanamlendea** [comment]
I have found another bug.

1. Enable the full Immunization module on the testing instance;
2. Create 2 duplicate Cases;
3. For each Case create an Immunization with Vaccination as Means of Immunization; Vaccination Date and Vaccine Name is identical for the two Vaccinations;
4. Merge the two Cases;
5. Open the Case and observe the Immunization Card;

-> The Case now has 2 Immunizations. The first one is the preserved Immunization (with the preserved Vaccination) and the second one is copied from the unpreserved Case; The second Immunization is empty because the Vaccination got merged into the preserved Case's Vaccination.

**Comment by MateStrysewske** [comment]
@roxanamlendea That's not a bug, but desired behaviour for now because this immunization could contain information that would otherwise be lost. We'll properly take care of it when introducing duplicate person merging.

**Comment by roxanamlendea** [comment]
Verified ticket on local environment using the latest version of Sormas 1.73.0-SNAPSHOT from the development branch.

**Comment by SahaLinaPrueger** [comment]
@roxanamlendea one question (last sentence): 

First example: 
Case 1 has 3 vaccinations (with Vaccination Date and Vaccine Name) and none of the three vaccinations has a _vaccination info source_ (fields are empty)
Case 2 has 3 vaccinations (with the same Vaccination Date and Vaccine Name as Case 1), the  fields _vaccination info source_  are not empty. 3x filled in with  _Vaccination card_.
Result after merging (Case 1 is leading case): 3 Vaccinations and the field  _vaccination info source_  is always filled in with _Vaccination card_. This is the expected result. 

Second example:
Case 1 has 3 vaccinations (with Vaccination Date and Vaccine Name). 
Vaccination 1: field _vaccination info source_  is empty
Vaccination 2: field _vaccination info source_  is filled in with _Vaccination card_
Vaccination 3: field _vaccination info source_  is empty

Case 2 has 3 vaccinations (with the same Vaccination Date and Vaccine Name as Case 1). 
Vaccination 1: field _vaccination info source_  is filled in with _oral communication_
Vaccination 2: field _vaccination info source_  is filled in with _oral communication_
Vaccination 3: field _vaccination info source_  is filled in with _oral communication_

Expected result after merging (Case 1 is leading case): 3 Vaccinations and the field  _vaccination info source_  is always filled (1x Vaccination Card, 2x oral communication)
Actual Result: after merging (Case 1 is leading case): 3 Vaccinations and the field  _vaccination info source_  is always filled (1x Vaccination Card, 1x oral communication, 1x empty/nothing)

The Question is: Why is the expected result in the second example not displayed?

**Comment by roxanamlendea** [comment]
@SahaLinaPrueger First of all, sorry for the late reply but I wanted to further investigate the second scenario and to figure our when it is reproducible and when it isn't.
As far as I can tell, the problem occurs when the following conditions are met:
- both Case 1 and Case 2 have multiple vaccinations (>1 vaccination each) 
- all of them have the same Vaccination Date and Vaccination Name
- the Vaccinations of the leading case have the 'Vaccination Info Source' set to null and the non-leading case has values set for the 'Vaccination Info Source'. 
The issue seems to be that only the first vaccination adopts the 'Vaccination Info Source'.
Another issue related to this is that before merging these Cases, there's an error message that is displayed.

![image](https://user-images.githubusercontent.com/89522688/177755854-d8e5e6f6-cae4-493b-a776-d45437cd0be3.png)

I will open a bug report for this, where I will link your comment and I will document there further scenarios.
